### ISSUE TITLE
View bookmarks by labels (group by labels)

### ISSUE THREAD

**Issue by mplebanski** [alefragnani/vscode-bookmarks#225]
Imagine you are working on a large codebase and you are trying to trace down a bug no. 1234. You mark some places in the code (using label bug 1234), finally fix the bug and wait for tests to finish. In the meantime you could start working on a refactor in some module X and you need to make some bookmarks again to understand how this module is used. Those bookmarks would be named 'PR1234'. 

Now this workflow results in bookmarks being intermingled with each other. 

I'd love if the new side bar had an option to show bookmarks by labels and then by files. Bookmarks with no label would go to a group 'Others',  'Unlabeled' or something like this.

**Comment by mplebanski** [comment]
Or even better would be to introduce some tags and leave labels untouched, because one would like to give a bookmark a meaningful name like: "This is where X is constructed".

**Comment by alefragnani** [comment]
Hi @mplebanski ,

> I'd love if the new side bar had an option to show bookmarks by labels and then by files. Bookmarks with no label would go to a group 'Others', 'Unlabeled' or something like this.

Do you mean _sorted ascending/descending by label_ or _grouped by similar labels_. Because in your example, `1234` is _similar_ to `PR1234` (if you think about a regex)? 

@mplebanski using tags would require adding _another concept_ in bookmarks, but that's a valid idea too.

Hope this helps

**Comment by mplebanski** [comment]
> Do you mean _sorted ascending/descending by label_ or _grouped by similar labels_. Because in your example, `1234` is _similar_ to `PR1234` (if you think about a regex)?

I meant grouping by label, but after a bit of thinking on  "sorting" I believe showing a sorted list of labels (sorted by name) is even better, since user can control the names, thus group them together in this list. Using sorted list one can even create logical subgroups by naming the labels right e.g.:

```
label_a
label_a_note
label_b
```

To sum up this feature request is about a new sorted labels view. one without grouping labels by files.

**Comment by alefragnani** [comment]
In this case, this issue could be _combined_ with #183, which asks for "stack" ordering of bookmarks. But, I would say _sort by ascending/descending_ is easier to achieve, because there is no need to store any additional info in the extension's data.

Hope this helps

**Comment by EmbeddedBacon** [comment]
I think the idea of group or categories would be very beneficial.  I have a number of bookmarker that I set over the course of months based upon that task.  I do have need to go back to those specific bookmarks when bugs are report, but it is hard to navigate though all the bookmarks I have created.  If the sidebar for Bookmarks had the ability to create groups/categories with the sidebar would nice.  By default when creating a new bookmark it would default to the group/category that is currently selected.  Would need the ability to drag/drop bookmarks from one group/category in case I make a mistake and group/categorize the bookmark which is like to happen more times than not.

**Comment by jotabe-net** [comment]
The full Visual Studio, not VS Code, has the concept of bookmark folders. In the Bookmarks window, equivalent of VS Code sidebar, you can create a hierarchy of folders, and put the bookmarks inside them, so it's really easy to manage the grouping. It's not compulsory to either create folders or put the bookmarks inside them, in other words, you can create "plain bookmarks in the root".
To support this it would be necessary to:
- support a syntax when creating the bookmar, something like "FolderName/SubfolderName/Named bookmark name"
- support the drag and drop, ant the tree view in the sidebar

It would look something like this:
```
Bookmark in the root
Other Bookmark in the root
-[Folder1]
  Bookmark in folder1
  -[Subfolder 1.1]
     Bookmark in subfolder 1.1
+[Folder2]
-[Folder3]
  +[Subfolder 3.1]
````

**Comment by phedlund113** [comment]
I can see the use of the sets or grouping of bookmarks. I have an issue i am tracking that i have 10 bookmarks for. I would find it nice to be able to group with a label so i can keep bookmarking with other issues and go back to previous issues when changes occur. I have tried the Label feature but that is not what i am after as far as bookmark organization.

The plugin is very useful as is so Im not throwing stones, just offering suggestions.

**Comment by phedlund113** [comment]
To clarify my request, in the sidebar the organization would be
Group1
    File1
       Bookmark1
       Bookmark2
    File2
       Bookmark1
    File3
       Bookmark1
       Bookmark2
       Bookmark3
Group2
    File1
       Bookmark1
       Bookmark2
Where Group name could be defined but default would be Group (or set)

**Comment by sharabiania** [comment]
I would like to see a feature to organize the bookmarks in groups (same way as folders) and see them in the side bar like this:


fake-group-1
   bookmark-fake-label-1
   bookmark-fake-label-2

fake-group-2
   bookmark-fake-label-2
   bookmark-fake-label-3

The same way we can organize the bookmarks in the Visual Studio.

One use case is, in large projects when I work on multiple tickets, I want to be able to categorize base on the Tickets:

ticket-id-1
  bookmark-1
  bookmark-2

...

**Comment by evgeny-goldin** [comment]
Found Favorites extension - https://marketplace.visualstudio.com/items?itemName=kdcro101.favorites. 
Not quite bookmarks, but lets one organize files into grouped favorites. 

Still following and +1-ing this thread as well!

**Comment by smallscript** [comment]
> Found Favorites extension - https://marketplace.visualstudio.com/items?itemName=kdcro101.favorites.
> Not quite bookmarks, but lets one organize files into grouped favorites.
> 
> Still following and +1-ing this thread as well!

I agree with the Visual Appeal and Interaction Patterns that the above "Favorites" extension offers. If it allowed me to Ctrl-K create bookmarks to files, then it would probably be my ideal answer.

**Comment by mihaibrana** [comment]
Became a patron to support adding this feature. Considering the number of users you have, I know adding a bit of wind in your sails will eventually trickle down to everybody! Go go Alessandro! :D 👯 

BTW: a good first start on this feature could simply be adding a button that shows ALL bookmarks without ANY grouping, sorted by label (or by path for those that don't have a label).

**Comment by alefragnani** [comment]
Thanks @mihaibrana ! I’ll be launching a poll in upcoming days, to have a better idea of which directions to take on upcoming releases, so it would be great if you add your vote/comment, when available.

About your suggestion, in the recent release you _amost_ have this. There is a new `View as List` option in the sidebar, which simply removes the `File` node, and displays all bookmarks in the root level. No, it doesn’t have any _sort_ yet, but it’s on the way.

**Comment by pltrant** [comment]
> Found Favorites extension - https://marketplace.visualstudio.com/items?itemName=kdcro101.favorites.
> Not quite bookmarks, but lets one organize files into grouped favorites.
> 
> Still following and +1-ing this thread as well!

Thanks for sharing this extension.  This is exactly what I need (the ability to bookmark/tag specific files, then sort them in a list with groups).  Ideally, I was hoping this was a feature of the Bookmarks extension itself, but the Favorites extension will work until the former can also gain similar functionality.  It boils down to needing the ability to easy tag files, then organize them in logical groups for easy review/recall.

**Comment by alefragnani** [comment]
About this feature in particular, I have a mixed feelings when compared to #254 .

I mean, if I add a **Tag** concept to bookmarks, similar to what I did with my Projects Manager extension (more details https://github.com/alefragnani/vscode-project-manager/issues/50#issuecomment-769053196), is it still necessary to have a _group by label_?

Sure, things like _Sort by Label_ option in Side Bar/Command Palette could still be available, just like #435 (Sort by Filename) will be in the future.

**Comment by atxsinn3r** [comment]
+1 for this feature.  To me grouping is important because I use bookmarks for noting bugs, and I need to categorize them in bug types, or some kind of group like "looks funny." Also when I triage them, I may need to move some of the bookmarks into another category named "Invalid". Basically I use bookmarks like you would with a browser's bookmarks.

**Comment by nkay08** [comment]
> About this feature in particular, I have a mixed feelings when compared to #254 .
> 
> I mean, if I add a **Tag** concept to bookmarks, similar to what I did with my Projects Manager extension (more details [alefragnani/vscode-project-manager#50 (comment)](https://github.com/alefragnani/vscode-project-manager/issues/50#issuecomment-769053196)), is it still necessary to have a _group by label_?
> 
> Sure, things like _Sort by Label_ option in Side Bar/Command Palette could still be available, just like #435 (Sort by Filename) will be in the future.

I also would like an option to manually group or tag my bookmarks, as I am working on different tasks.

The way the tags work in your Projects Manager is good and translating that feature to the Bookmarks extension would be great and very helpful.

**Comment by ckx** [comment]
I would also very much appreciate grouping. My current workaround is to do my own tagging in the bookmark label:

![image](https://github.com/alefragnani/vscode-bookmarks/assets/1069665/596b8bab-2df6-4add-957f-5c5ebb7ae684)

I don't like this very much, but it's better than nothing I guess. Maybe a search/filter would be helpful in combination with this approach.

**Comment by alefragnani** [comment]
Hi,

With a new year about to start, it is the moment to review the extension’s backlog in order to keep it healthy and focused.

I'm closing this issue as a duplicate of#254, as it shares most of the main ideas behind it, as commented here (https://github.com/alefragnani/vscode-bookmarks/issues/225#issuecomment-914505046).

Feel free to upvote and/or share comments there.

Thanks for your understanding
